src/OpenLoco/Windows/TerraForm.cpp

```diff
@@ -395,27 +408,297 @@ namespace OpenLoco::Ui::Windows::Terraform
+        }
+
+        // 0x004BDC67
+        static void clusterSurfaceTypeTreeToolDown(const GameCommands::TreePlacementArgs& baseArgs, const uint16_t range, const uint16_t density)
+        {
+            const auto numPlacements = (range * range * density) / 8192;
+            for (auto i = 0; i < numPlacements; ++i)
+            {
+                // Choose a random offset in a circle
+                auto& rng = gPrng();
+                auto randomMagnitude = rng.randNext(std::numeric_limits<uint16_t>::max()) * range / 65536;
+                auto randomDirection = rng.randNext(Math::Trigonometry::directionPrecisionHigh - 1);
+                Map::Pos2 randomOffset(
+                    Math::Trigonometry::integerSinePrecisionHigh(randomDirection, randomMagnitude),
+                    Math::Trigonometry::integerCosinePrecisionHigh(randomDirection, randomMagnitude));
+
+                GameCommands::TreePlacementArgs args;
+                Map::Pos2 newLoc = randomOffset + baseArgs.pos;
+                args.quadrant = ViewportInteraction::getQuadrantFromPos(newLoc);
+                args.pos = Map::Pos2(newLoc.x & 0xFFE0, newLoc.y & 0xFFE0);
+                // Note: this is not the same as the randomDirection above as it is the trees rotation
+                args.rotation = rng.randNext(3);
+                args.colour = 0;
+                auto type = getRandomTreeTypeFromSurface(newLoc, false);
+                if (type)
+                {
+                    args.type = *type;
+                    args.buildImmediately = true;
+                    args.requiresFullClearance = true;
+                    do_23(GameCommands::Flags::apply, args);
+                }
+            }
         }
 
         // 0x004BBB20
         static void onToolDown(Window& self, const WidgetIndex_t widgetIndex, const int16_t x, const int16_t y)
         {
-            registers regs;
-            regs.esi = X86Pointer(&self);
-            regs.dx = widgetIndex;
-            regs.ax = x;
-            regs.bx = y;
-            call(0x004BBB20, regs);
+            removeTreeGhost();
+            auto placementArgs = getTreePlacementArgsFromCursor(x, y);
+            if (placementArgs)
+            {
+                switch (_treeClusterType)
+                {
+                    case treeCluster::none:
+                        if (GameCommands::do_23(GameCommands::Flags::apply, *placementArgs) != GameCommands::FAILURE)
+                        {
+                            Audio::playSound(Audio::SoundId::construct, GameCommands::getPosition());
+                        }
+                        break;
+                    case treeCluster::selected:
+                    {
+                        // TODO: Lock behind sandbox/editor
+                        auto previousId = CompanyManager::updatingCompanyId();
+                        CompanyManager::updatingCompanyId(CompanyId::neutral);
```

Comment:
This is great. Looks like it won't be that much work to make tree clusters cost money, then. Something you'd prefer to tackle in another PR?

Reply:
Yeah i think so